### PR TITLE
switch to new rowan API

### PR DESCRIPTION
**Pull request by matklad**
closes https://github.com/rust-analyzer/rust-analyzer/issues/448

### CODE REVIEW

**Comment by matklad** [comment]
Corresponing rowan PR: https://github.com/rust-analyzer/rowan/pull/9

**Comment by matklad** [comment]
@flodiebold I would appreciate a reivew of this and, especially, the rowan PR :-) 

It would be sad if there's some nasty [safety] hole in the proposed API in the end...

**Comment by flodiebold** [comment]
I'm not that experienced with unsafe, so I don't know how much I can help there, but I'll take a look nonetheless ;)

File: crates/ra_syntax/src/ast.rs [hunk 1]
```diff
@@ -265,90 +261,90 @@ impl<'a> Whitespace<'a> {
     }
 }
```
Comment (csmoe) [inline crates/ra_syntax/src/ast.rs]: how about wrapping those impl_test() into macro?

File: crates/ra_syntax/src/ast/generated.rs [hunk 2]
```diff
@@ -9,4648 +9,3387 @@
...
-
-impl<R: TreeRoot<RaTypes>> EnumVariantListNode<R> {
-    pub fn borrowed(&self) -> EnumVariantList {
-        EnumVariantListNode { syntax: self.syntax.borrowed() }
-    }
-    pub fn owned(&self) -> EnumVariantListNode {
-        EnumVariantListNode { syntax: self.syntax.owned() }
-    }
+    fn syntax(&self) -> &SyntaxNode { &self.syntax }
+    fn to_owned(&self) -> TreePtr<EnumVariantList> { TreePtr::cast(self.syntax.to_owned()) }
```
Comment (csmoe) [inline crates/ra_syntax/src/ast/generated.rs]: and those `AstNode` implementations too with proc_macro.

File: crates/ra_syntax/src/ast.rs (hunk 1 above)
Comment (matklad) [reply crates/ra_syntax/src/ast.rs]: Yeah, I'll plan to code generate them soonish.

File: crates/ra_syntax/src/ast/generated.rs (hunk 2 above)
Comment (matklad) [reply crates/ra_syntax/src/ast/generated.rs]: These all are already generated from a `grammar.ron` with a script. I feel that a good old code generation would work better than a proc macro here:

* easy to understand, bc you see the generated code
* fast to compile (you don't generate the impls on every compilation)
* easy to modify (template is just a tera file)

Comment (csmoe) [reply crates/ra_syntax/src/ast/generated.rs]: oops, my fault, didn't notice the `generated.rs` name.

**Pull request merged**